feat(ecs-patterns): support FargatePlatformVersion as an input property

michaelmoussa opened the pull request:
### Commit Message
Allow setting Fargate Platform version with all of the LoadBalancedFargateServices and QueueProcessingService.

Closes #7476 
### End Commit Message
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 2906e99f57bddb0cc3c8f41baddfbaa60c9d9c9e
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=91c92353-e84a-45c1-bf88-82f4c159a536%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 4856a7cea684170cdaf7c07e98b1e3a84aa62191
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=704c49ea-a617-4ac0-8054-901218822856%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

pkandasamy91 commented on packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.load-balanced-fargate-service-v2.ts:
```diff
@@ -4,6 +4,7 @@ import { AwsLogDriver, Cluster, ContainerImage, FargateTaskDefinition, Propagate
 import { Duration, Stack } from '@aws-cdk/core';
 import { Test } from 'nodeunit';
+import {FargatePlatformVersion} from '../../../aws-ecs/lib/fargate/fargate-service';
 import { ApplicationMultipleTargetGroupsFargateService, NetworkMultipleTargetGroupsFargateService } from '../../lib';
```
You already have the `@aws-cdk/aws-ecs` module imported, you should add it above 

update 
`import { AwsLogDriver, Cluster, ContainerImage, FargateTaskDefinition, PropagatedTagSource, Protocol } from '@aws-cdk/aws-ecs';` 
to 
`import { AwsLogDriver, Cluster, ContainerImage, FargatePlatformVersion, FargateTaskDefinition, PropagatedTagSource, Protocol } from '@aws-cdk/aws-ecs';`

pkandasamy91 commented on packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.load-balanced-fargate-service.ts:
```diff
@@ -5,6 +5,7 @@ import { ApplicationLoadBalancer, ApplicationProtocol, NetworkLoadBalancer } fro
 import * as iam from '@aws-cdk/aws-iam';
 import * as cdk from '@aws-cdk/core';
 import { Test } from 'nodeunit';
+import {FargatePlatformVersion} from '../../../aws-ecs/lib/fargate/fargate-service';
```
Since you're already importing the entire ecs module, you can refer to it with ecs.FargatePlatformVersion

pkandasamy91 commented on packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.load-balanced-fargate-service.ts:
```diff
@@ -228,6 +229,24 @@ export = {
     test.done();
   },
 
+  'setting platform version'(test: Test) {
```
We have a test called `'test Fargate loadbalanced construct with all settings'` in which we specify every supported property. Updating that test should be sufficient.

michaelmoussa replied:
That test is in a different file for a different construct (`test.load-balanced-fargate-service-v2.ts`). I did use the test for every property in that one, but I didn't see anything like that in this file.

pkandasamy91 replied:
If you can add the test in that file, that would be great! (also if you can add a test for all 4 load balanced service constructs that would be awesome!)

pkandasamy91 commented on packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.load-balanced-fargate-service-v2.ts:
```diff
@@ -121,6 +122,7 @@ export = {
         desiredCount: 3,
         enableECSManagedTags: true,
         healthCheckGracePeriod: Duration.millis(2000),
+        platformVersion: FargatePlatformVersion.VERSION1_4,
```
You also want to add this to the expect section of the test

pkandasamy91 commented on packages/@aws-cdk/aws-ecs-patterns/test/fargate/test.queue-processing-fargate-service.ts:
```diff
@@ -4,6 +4,7 @@ import * as ecs from '@aws-cdk/aws-ecs';
 import * as sqs from '@aws-cdk/aws-sqs';
 import * as cdk from '@aws-cdk/core';
 import { Test } from 'nodeunit';
+import {FargatePlatformVersion} from '../../../aws-ecs/lib/fargate/fargate-service';
```
Same comment as above.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: e1046d3b00c79230cc1f1005ed11e71728743c76
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fc75f0e3-b479-4ce1-8624-6c03ad90ed28%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

michaelmoussa commented on packages/@aws-cdk/aws-ecs-patterns/lib/fargate/application-load-balanced-fargate-service.ts:
```diff
@@ -64,6 +65,17 @@ export interface ApplicationLoadBalancedFargateServiceProps extends ApplicationL
+   * [AWS Fargate Platform Versions](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/platform_versions.html)
+   * in the Amazon Elastic Container Service Developer Guide.
+   *
+   * @default Latest
```
I was following the existing convention [here](https://github.com/aws/aws-cdk/blob/1ee629e418fccec30b8a94e43682ed2c47ddd8da/packages/%40aws-cdk/aws-ecs/lib/fargate/fargate-service.ts#L49) in the `FargateService`.

Should we update it there as well?

pkandasamy91 replied:
Ah I see, for the time being, I would say leave it as is, and if it becomes a point of contention, we can revisit updating it.

michaelmoussa commented:
@pkandasamy91 I've removed the redundant `import`s and also added the missing test expectation. For the `@default - Latest` vs. `@default none`, I feel we should also update the `FargateService` accordingly, and that we need to update the comment so that it doesn't conflict with the `@default none`. 

They currently say:

```javascript
  /**
   * The platform version on which to run your service.
   *
   * If one is not specified, the LATEST platform version is used by default. For more information, see
   * [AWS Fargate Platform Versions](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/platform_versions.html)
   * in the Amazon Elastic Container Service Developer Guide.
   *
   * @default Latest
   */
  readonly platformVersion?: FargatePlatformVersion;
```

Perhaps we should change it to:

```javascript
  /**
   * The platform version on which to run your service.
   *
   * If one is not specified, Fargate itself will set the LATEST platform version by default. For more information, see
   * [AWS Fargate Platform Versions](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/platform_versions.html)
   * in the Amazon Elastic Container Service Developer Guide.
   *
   * @default none
   */
  readonly platformVersion?: FargatePlatformVersion;
```

What do you think?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: ad739d23268fb1f8d757fccb0cf383381e53a86a
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e683e474-ac04-46a3-a1ff-5cb097a9accf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b1dc626f19fe179b1075e20145d4bfb36803740a
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fff7a826-34d3-48c8-a539-e4102071906b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: fcf70175f9921b84bb3a915523866ce8f9c826e6
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bc4e9724-fe9e-4aac-97c6-278df0220464%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
